### PR TITLE
Issue 943

### PR DESCRIPTION
**Pull request by laurentder**

### CODE REVIEW

File: cargo-creusot/src/why3_launcher.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        match self {
+            Why3Mode::Ide => f.write_str("ide"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Why3Launcher {
+    mode: Why3Mode,
+    why3_path: Option<PathBuf>,
```
Comment (xldenis) [inline cargo-creusot/src/why3_launcher.rs:28]: Given that we cannot launch why3 (according to line 47) without this, can we make this not optional? To me, the error should be launched in the builder instead.

File: cargo-creusot/src/why3_launcher.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    ) -> Self {
+        Self { mode, why3_path, config_file, args, output_file }
+    }
+
+    pub fn make(&self, temp_dir: &Path) -> Result<Command> {
+        let Some(ref mlcfg_file) = self.output_file else {
+            return Err(anyhow!("can't launch why3, no mlcfg_file specify"));
+        };
+        let mode = self.mode.to_string();
+        PRELUDE.extract(temp_dir).expect("can't launch why3, could extract prelude into temp dir");
```
Comment (xldenis) [inline cargo-creusot/src/why3_launcher.rs:50]: Not that I have any objection with this, but why is it now necessary to copy the prelude to a temporary directory? Just hygiene? Or is there a more fundamental reason?

**Review by xldenis — COMMENTED** [review]
Thanks for this work! I think it's a nice step forward. I think a follow up could involve removing the code in `creusot-rustc` which is made obsolete by this, though I think that would require also handling `prove` and `replay`. cc @dewert99.

File: cargo-creusot/src/why3_launcher.rs (hunk 2 above)
Comment (laurentder) [reply cargo-creusot/src/why3_launcher.rs:50]: I took this code in: creusot/src/run_why3.rs::run_why3

File: cargo-creusot/src/why3_launcher.rs (hunk 1 above)
Comment (laurentder) [reply cargo-creusot/src/why3_launcher.rs:28]: Ok, like you want

**Comment by laurentder** [comment]
> Thanks for this work! I think it's a nice step forward. I think a follow up could involve removing the code in `creusot-rustc` which is made obsolete by this, though I think that would require also handling `prove` and `replay`. cc @dewert99.

yes the final goal is also to manage the "prove" and the "replay", however I will not have time this week to achieve this.

**Comment by xldenis** [comment]
> yes the final goal is also to manage the "prove" and the "replay", however I will not have time this week to achieve this.

That's totally fine, but I am looking forward to the day when we can clean that up :)

**Pull request merged**